tests/returntypes/test_vjp_new.py

```diff
@@ -0,0 +1,665 @@
+
+    def test_compute_multiple_measurement_multi_dim_multiple_params(self):
+        """Test that the correct VJP is returned"""
+        dy = np.array([[1.0, 2.0], [3.0, 4.0]])
+        jac = np.array([[[1.0, 0.1, 0.2], [0.2, 0.6, 0.1]], [[0.4, -0.7, 1.2], [-0.5, -0.6, 0.7]]])
+
+        vjp = qml.gradients.compute_vjp(dy, jac)
+
+        assert vjp.shape == (3,)
+        assert np.all(vjp == np.tensordot(dy, jac, axes=[[0, 1], [0, 1]]))
+
+    def test_jacobian_is_none_single(self):
+        """A None Jacobian returns a None VJP"""
+
+        dy = np.array([[1.0, 2.0], [3.0, 4.0]])
+        jac = None
+
+        vjp = qml.gradients.compute_vjp_single(dy, jac)
+        assert vjp is None
+
+    def test_jacobian_is_none_multi(self):
+        """A None Jacobian returns a None VJP"""
+
+        dy = np.array([[1.0, 2.0], [3.0, 4.0]])
+        jac = None
+
+        vjp = qml.gradients.compute_vjp_multi(dy, jac)
+        assert vjp is None
+
+    def test_zero_dy_single_measurement_single_params(self):
+        """A zero dy vector will return a zero matrix"""
+        dy = np.zeros([1])
+        jac = np.array(0.1)
+
+        vjp = qml.gradients.compute_vjp_single(dy, jac)
+        assert np.all(vjp == np.zeros([1]))
+
+    def test_zero_dy_single_measurement_multi_params(self):
+        """A zero dy vector will return a zero matrix"""
+        dy = np.zeros([2])
+        jac = tuple([np.array(0.1), np.array(0.2)])
+
+        vjp = qml.gradients.compute_vjp_single(dy, jac)
+        assert np.all(vjp == np.zeros([2]))
+
+    def test_zero_dy_multi(self):
+        """A zero dy vector will return a zero matrix"""
+        dy = tuple([np.array(0.0), np.array([0.0, 0.0])])
+        jac = tuple(
+            [
+                tuple([np.array(0.1), np.array(0.1), np.array(0.1)]),
+                tuple([np.array([0.1, 0.2]), np.array([0.1, 0.2]), np.array([0.1, 0.2])]),
+            ]
+        )
+
+        vjp = qml.gradients.compute_vjp_multi(dy, jac)
+        assert np.all(vjp == np.zeros([3]))
+
+    @pytest.mark.torch
+    @pytest.mark.parametrize("dtype1,dtype2", [("float32", "float64"), ("float64", "float32")])
+    def test_dtype_torch(self, dtype1, dtype2):
+        """Test that using the Torch interface the dtype of the result is
+        determined by the dtype of the dy."""
+        import torch
+
+        dtype1 = getattr(torch, dtype1)
+        dtype2 = getattr(torch, dtype2)
+        a = torch.ones((1), dtype=dtype1)
+        b = torch.ones((2), dtype=dtype1)
+
+        dy = tuple([a, b])
+        jac = tuple([torch.ones((1), dtype=dtype2), torch.ones((2), dtype=dtype2)])
+
+        assert qml.gradients.compute_vjp_multi(dy, jac)[0].dtype == dtype1
+
+    @pytest.mark.tf
+    @pytest.mark.parametrize("dtype1,dtype2", [("float32", "float64"), ("float64", "float32")])
+    def test_dtype_tf(self, dtype1, dtype2):
+        """Test that using the TensorFlow interface the dtype of the result is
+        determined by the dtype of the dy."""
```

Comment:
The user has no control over `dy` iirc. Does that mean they won't be able to control the dtype of jacobians?